### PR TITLE
Added checks for pages specified in 'error_page' and 'site_unavailable_page' settings

### PR DESCRIPTION
**Pull request by Ruslan-Aleev**
### What does it do?
The resource that is specified in the system settings as `site_start` cannot be unpublished from the context menu in tree - an error will appear. But at the same time, this resource can be unpublished trough chekbox in resource form.

The resources specified in `error_page` and `site_unavailable_page` could be deleted / unpublished both through the tree and through the resource form.

**In general, I added checks for these resources.**

### Related issue(s)/PR(s)
https://github.com/modxcms/revolution/issues/14509

### CODE REVIEW

**Comment by Ruslan-Aleev** [comment]
In the process of working on PR, strange settings key are annoying! :)

I would like to have more logical names:
```
site_index (instead of site_start);
site_404 (instead of error_page);
site_503 (instead of site_unavailable_page).
```

**Comment by ghost** [comment]
What about resource fields and database field names, like "introtext" and "summary"? I nagged about that one in the beginning until (introtext) was added next to the field label.

**Comment by Ruslan-Aleev** [comment]
@sottwell Yes, it looks strange too.

**Review by Ibochkarev — APPROVED** [review]
изображение

изображение

изображение

изображение

File: core/src/Revolution/Processors/Resource/Update.php [hunk 1]
```diff
@@ -479,18 +477,48 @@ public function checkPublishingPermissions()
      *
      * @return boolean
      */
-    public function checkForUnPublishOnSiteStart()
+    public function checkForUnPublishOnSiteSettingPages()
     {
-        $passed = true;
+        $result = true;
```
Comment (theboxer) [inline core/src/Revolution/Processors/Resource/Update.php]: Unnecessary variable

File: core/src/Revolution/Processors/Resource/Update.php [hunk 2]
```diff
@@ -479,18 +477,48 @@ public function checkPublishingPermissions()
...
+            $result = $this->modx->lexicon('resource_err_unpublish_errorpage_dates');
+        } else if ($isSiteUnavailablePage && $deleted == 1) {
+            $result = $this->modx->lexicon('resource_err_delete_siteunavailable');
+        } else if ($isSiteUnavailablePage && $published == 0) {
+            $result = $this->modx->lexicon('resource_err_unpublish_siteunavailable');
+        } else if ($isSiteUnavailablePage && (!empty($publishDate) || !empty($unPublishDate))) {
+            $result = $this->modx->lexicon('resource_err_unpublish_siteunavailable_dates');
+        }
+
+        return $result;
```
Comment (theboxer) [inline core/src/Revolution/Processors/Resource/Update.php]: Change to `return true;`

File: core/src/Revolution/Processors/Resource/Update.php [hunk 3]
```diff
@@ -479,18 +477,48 @@ public function checkPublishingPermissions()
...
+        }
+        if ($this->getProperty('id') == $workingContext->getOption('error_page') || $this->getProperty('id') == $this->modx->getOption('error_page')) {
+            $isSiteErrorPage = true;
+        }
+        if ($this->getProperty('id') == $workingContext->getOption('site_unavailable_page') || $this->getProperty('id') == $this->modx->getOption('site_unavailable_page')) {
+            $isSiteUnavailablePage = true;
+        }
+
+        if ($isSiteStart && $deleted == 1) {
+            $result = $this->modx->lexicon('resource_err_delete_sitestart');
```
Comment (theboxer) [inline core/src/Revolution/Processors/Resource/Update.php]: Follow return early pattern.

replace with `return $this->modx...`

File: core/src/Revolution/Processors/Resource/Update.php [hunk 4]
```diff
@@ -479,18 +477,48 @@ public function checkPublishingPermissions()
...
+        } else if ($isSiteStart && $published == 0) {
```
Comment (theboxer) [inline core/src/Revolution/Processors/Resource/Update.php]: Replace all `else if` with a regular `if` and follow return early pattern.

`return $this->modx...`

File: core/src/Revolution/Processors/Resource/Delete.php [hunk 5]
```diff
@@ -120,6 +123,36 @@ public function process()
...
+
+    /**
+     * Checks if the given resource is set as site_unavailable_page
+     * @return bool
+     */
+    public function isSiteUnavailablePage()
+    {
+        $workingContext = $this->modx->getContext($this->getProperty('context_key', $this->resource->get('context_key') ? $this->resource->get('context_key') : 'web'));
+        return ($this->resource->get('id') == $workingContext->getOption('site_unavailable_page') || $this->resource->get('id') == $this->modx->getOption('site_unavailable_page'));
+    }
```
Comment (theboxer) [inline core/src/Revolution/Processors/Resource/Delete.php]: This calls for some abstraction as you are redefining these checks 3 times.

**Review by theboxer — CHANGES_REQUESTED** [review]

**Comment by Ruslan-Aleev** [comment]
@theboxer I have corrected the code. Any thoughts?

**Comment by Ruslan-Aleev** [comment]
@theboxer Please check PR.

**Review by theboxer — APPROVED** [review]

**Comment by theboxer** [comment]
Seems good to me, thanks!

**Review by opengeek — APPROVED** [review]

**Pull request merged**